README.md

JdeRobot
========

<!--[![Build Status](http://jenkins.jderobot.org/buildStatus/icon?job=JdeRobot-GitHub)](http://jenkins.jderobot.org/job/JdeRobot-GitHub/)-->
[![Current Release Version](https://img.shields.io/github/release/jderobot/base?style=flat-square&logo=github)](https://github.com/explosion/spaCy/releases)

# Introduction


JdeRobot is a software development suite for robotics, home-automation and computer vision applications. These domains include sensors (for instance, cameras), actuators, and intelligent software in between. It has been designed to help in programming with such intelligent softwares. It is mainly written in C++ language and provides a distributed component-based programming environment where the application program is made up of a collection of several concurrent asynchronous components. Each component may run in different computers and they are connected using Ice communication middleware. Components may be written in C++, python, Java... and all of them interoperate through explicit Ice interfaces.

JdeRobot simplifies the access to hardware devices from the control program. Getting sensor measurements is as simple as calling a local function, and ordering motor commands as easy as calling another local function. The platform attaches those calls to the remote invocation on the components connected to the sensor or the actuator devices. They can be connected to real sensors and actuators or simulated ones, both locally or remotely using the network. Those functions build the API for the Hardware Abstraction Layer. The robotic application get the sensor readings and order the actuator commands using it to unfold its behavior. Several driver components have been developed to support different physical sensors, actuators and simulators. The drivers are used as components installed at will depending on your configuration. They are included in the official release. Currently supported robots and devices:

* RGBD sensors: Kinect from Microsoft, Asus Xtion
* Pioneer robot from MobileRobotics Inc.
* Kobuki robot (TurtleBot) from Yujin Robot
* Nao humanoid from Aldebaran
* ArDrone quadrotor from Parrot
* Firewire cameras, USB cameras, video files (mpeg, avi...), IP cameras (like Axis)
* Pantilt unit PTU-D46 from Directed Perception Inc.
* Laser Scanners: LMS from SICK and URG from Hokuyo
* EVI PTZ camera from Sony
* Gazebo and Stage simulators
* Wiimote
* X10 home automation devices

JdeRobot includes several robot programming tools and libraries. First, viewers and teleoperators for several robots, its sensors and motors. Second, a camara calibration component and a tunning tool for color filters. Third, VisualStates tool for programming robot behavior using hierarchical finite state machines. It includes many sample components using OpenCV, PCL, OpenGL, etc.. In addition, it also provides a library to develop fuzzy controllers, a library for projective geometry and some computer vision processing.

Each component may have its own independent Graphical User Interface or none at all. Currently, GTK and Qt libraries are supported, and several examples of OpenGL for 3D graphics with both libraries are included.

JdeRobot is open-source software, licensed as GPL and LGPL. It also uses third-party softwares like Gazebo simulator, OpenGL, GTK, Qt, Player, Stage, Gazebo, GSL, OpenCV, PCL, Eigen, Ogre.

JdeRobot is a project developed by Robotics Group of Universidad Rey Juan Carlos (Madrid, Spain).

* Official Web Page: http://jderobot.org

# Installation

## Table of Contents

* [Getting environment ready](#env )
* [Installing from debian packages](#deb)
  * [Updating pakcages](#upd)
* [Installing from docker image](#dim)
* [Manual installation](#manual)
  * [Dependencies](#deps)
  * [Install from source](#source)



<a name="env"/>

### Getting environment ready

* Add the lastest ROS sources:

```
sudo sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
sudo apt-key adv --keyserver 'hkp://keyserver.ubuntu.com:80' --recv-key C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654
```

* Add the lastest Gazebo sources:

```
sudo sh -c 'echo "deb http://packages.osrfoundation.org/gazebo/ubuntu-stable `lsb_release -cs` main" > /etc/apt/sources.list.d/gazebo-stable.list'
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-key 67170598AF249743
```

* Add the lastest zeroc-ice sources:

```
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv B6391CB2CFBA643D
sudo apt-add-repository "deb http://zeroc.com/download/Ice/3.7/ubuntu18.04 stable main"
```
* Add JdeRobot repository (using dedicated file /etc/apt/sources.list.d/jderobot.list):

```
sudo sh -c 'cat<<EOF>/etc/apt/sources.list.d/jderobot.list

deb [arch=amd64] http://wiki.jderobot.org/apt bionic main
EOF'
```

* Get and add the public key from the JdeRobot repository

```
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv 24E521A4
```

* Update the repositories
```
sudo apt update
```



<a name="deb"/>
### Installing from debian packages (RECOMMENDED)

* Install JdeRobot:

```
sudo apt install jderobot
sudo apt install jderobot-assets
```

* After installing the package, you can close the terminal and reopen it to source the environment variables, OR just type: 

```
source ~/.bashrc
```
<a name="upd"/>

#### Updating packages

* If you already have a previous version of the packages installed, you only have to do:
```
sudo apt update && sudo apt upgrade
```



<a name="dim"/>

### Installing from docker image

If you want to run JdeRobot in MS-Windows, MacOS or other Linux distributions you can use Docker containers. We have created a [Docker image](https://www.docker.com/) with current JdeRobot Release and all the necessary components to be used with [JdeRobot-Academy](http://wiki.jderobot.org/JdeRobot-Academy). To download it, use: 

```
docker pull jderobot/jderobot
```

For more information follow [this link](https://hub.docker.com/r/jderobot/jderobot/)  


<a name="manual"/>

### Manual installation

Downloading the source code from the GitHub is strongly NOT RECOMMENDED for new users unless you know what you are doing. 

<a name="deps"/>

#### Dependencies

You have two options here:
1. Install all the dependencies from a binary package
2. Install all the dependencies manually  (**NOT RECOMMENDED**)

For first one, you only have to type the following:

```
sudo apt install jderobot-deps-dev
```
and skip to the next section of this README.

For the second one... keep reading

JdeRobot has different external dependencies to build its structure. There are two types of dependencies: necessary dependencies and other dependencies. The firsts are needed to compile and install the basics of JdeRobot, that is, all its libraries and interfaces that are needed for the components or to develop components that use JdeRobot. The seconds are needed just for some components, but are not really necessary unless you want to use that components. For instance, the component gazeboserver uses Gazebo, a 3D robot simulator, so to use this component you may install Gazebo first.

First of all repeat the step [Getting environment ready](#env )

Some libraries are required to compile, link or run JdeRobot. Just type the following commands:

* **Basic libraries:**

```sudo apt install build-essential libtool cmake g++ gcc git make```

* **OpenGL libraries:**

```sudo apt install freeglut3 freeglut3-dev libgl1-mesa-dev libglu1-mesa```

* **GTK2 libraries:**

```
sudo apt install libgtk2.0-0 libgtk2.0-bin libgtk2.0-cil libgtk2.0-common libgtk2.0-dev libgtkgl2.0-1
sudo apt install libgtkgl2.0-dev libgtkglext1 libgtkglext1-dev libglademm-2.4-dev libgtkmm-2.4-dev 
sudo apt install libgnomecanvas2-0 libgnomecanvas2-dev  libgtkglext1-doc libgnomecanvasmm-2.6-dev
sudo apt install libgnomecanvasmm-2.6-1v5 libgtkglextmm-x11-1.2-0v5 libgtkglextmm-x11-1.2-dev
```

* **Gtk3 libraries:**

```sudo apt install libgoocanvasmm-2.0-6 libgoocanvasmm-2.0-dev```

* **GSL libraries:**

```sudo apt install libgsl23 gsl-bin libgsl-dev```

* **LibXML:**

```sudo apt install libxml++2.6-2v5 libxml++2.6-dev libtinyxml-dev```

* **Eigen:**

```sudo apt install libeigen3-dev```

* **Fireware:**

```sudo apt install libdc1394-22 libdc1394-22-dev```

* **USB:**

```sudo apt install libusb-1.0-0 libusb-1.0-0-dev```

* **CWIID:**

```sudo apt install libcwiid-dev```

* **Python components:**

```sudo apt install python-matplotlib python-pyqt5 python-pip python-numpy python-pyqt5.qtsvg```

* **Qfi**

It can be compiled and installed from source: https://github.com/JdeRobot/ThirdParty/tree/master/qflightinstruments

* **Qt 5**

```sudo apt install qtbase5-dev libqt5script5 libqt5svg5-dev```

* **Boost**

```sudo apt install libboost-system-dev libboost-filesystem-dev```

* **ROS**

```
sudo apt install ros-melodic-roscpp ros-melodic-std-msgs ros-melodic-cv-bridge ros-melodic-image-transport ros-melodic-roscpp-core ros-melodic-rospy ros-melodic-nav-msgs ros-melodic-geometry-msgs ros-melodic-mavros ros-melodic-gazebo-plugins
```

Once all ros packages are installed, install the script that tunes the environment variables ROS in your .bashrc configuration file, and run it for the current shell:

```
echo "source /opt/ros/melodic/setup.bash" >> ~/.bashrc 
source ~/.bashrc 
```

* **Google glog (logging)**

```sudo apt install libgoogle-glog-dev```

* **GStreamer**

```sudo apt-get install libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev```

* **ICE**

```sudo apt install libdb5.3-dev libdb5.3++-dev libssl-dev libbz2-dev libmcpp-dev```

compile ice:

```
git clone -b 3.7 https://github.com/zeroc-ice/ice.git 
cd ice/cpp
make CPP11=yes OPTIMIZE=yes
make install
```
Configure ICE for Python with pip
```
sudo pip2 install --upgrade pip
sudo pip2 install zeroc-ice==3.7.2
```

* **OpenNI 2**

```sudo apt-get install libopenni2-dev libopenni-dev```

* **Point Cloud Library**

```sudo apt-get install libpcl-dev```

* **OpenCV**

```sudo apt-get install libopencv-dev```

* **NodeJS**

```sudo apt-get install nodejs```

* **Kobuki robot libraries**

You can find the source code in our git repository (http://github.com/jderobot/thirdparty.git)

* **SDK Parrot for ArDrone**

If you want to install it manually from our third party repository, you only have to:

1. Create a folder to compile the code
```mkdir ardronelib-build && cd ardronelib-build```

2. Download the installer, a CMakeLists.txt file
```
wget https://raw.githubusercontent.com/RoboticsURJC/JdeRobot-ThirdParty/master/ardronelib/CMakeLists.txt
wget https://raw.githubusercontent.com/RoboticsURJC/JdeRobot-ThirdParty/master/ardronelib/ffmpeg-0.8.pc.in
wget https://raw.githubusercontent.com/RoboticsURJC/JdeRobot-ThirdParty/master/ardronelib/ardronelib.pc.in
```
3. Compile and install as usual:
```
cmake .
make
sudo make install
```

<a href="#source"/>

#### Install from source

After installing all the dependencies you can compile the project, you can clone this repo and build it doing the following:

-  Download the source code from git:

```
git clone http://github.com/RoboticsURJC/JdeRobot.git
cd JdeRobot/
```

- Check system and dependencies

```
mkdir build && cd build
cmake ..
```

- Compile

```
make
```

- Install

```
sudo make install
```

# How To Contribute
To see the collaborate workflow and coding style of `JdeRobot` community, please refer to the [wiki page](https://github.com/RoboticsURJC/JdeRobot/wiki/How-To-Contribute).

# Copyright and license


    Copyright 2015 - JderRobot Developers
    
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
=======


debian/changelog
jderobot (6.1.0) unstable; urgency=low

  * release 6.1.0
  * Major cleanup of the repo. Unused or moved tools and drivers are removed. If you want to rescue them, go to tag 6.0.0
  * Fix compilation issues
  * Fix packages dependencies and minor ROS version issues.

-- Francisco Perez <f.perez475@gmail.com>, Tue, 12 Mar 2019 13:00:00 +0100

jderobot (5.6.7) unstable; urgency=low

  * release 5.6.7
  * Solved some bugs related with installation and other minor bugs in components.
  * [PR #1312] Unified Pibot model, now it has all its sensors active.
  * [PR #1313] Added support for MavRos, added iris drone and gazebo infraestructure for MavROS drones
  * [PR #1317] Refactorization of PiBot real driver, added odometry and documentation.
  * [PR #1321] Added printable pieces of Pibot in freecad format
  * [PR #1323] Solved texture loading of F1 cars for gzweb
  * [PR #1326] Added support for real turtlebot 
  * [PR #1328] Little refactorization of PiBot real driver
  * [PR #1338] Added a refactored picam printable piece 

-- Francisco Perez <f.perez475@gmail.com>, Tue, 12 Mar 2019 13:00:00 +0100

jderobot (5.6.6) unstable; urgency=low

  * release 5.6.6
  * Solved some bugs related with installation and other minor bugs in components.
  * [PR #1088] New PiCameraServer Driver and fixed Logging of slice executables error cmake
  * [PR #1156] Solved problem with usbcam on camserver_web
  * [PR #1167] Modify visualization.ice to visualize 3D segments
  * [PR #1169] 3DViewer_web modify to visualize 3D segments
  * [PR #1177] Added Roomba robot for ros nodes
  * [PR #1179] Added library PiBot as a new driver of JdeRobot
  * [PR #1180] Added driver piCamServer_py to serve PiCamera and USB cams.
  * [PR #1181] Added new tool PiCamCalibrator
  * [PR #1192] Added pibot robot support for gazebo
  * [PR #1198] Moved kids interfaces to interfaces directory
  * [PR #1209] New F1 model with realistic mesh (obj)
  * [PR #1220] Added model pibot with us sensor instead of picam sensor 
  * [PR #1228] New ROS-based worlds for vacuum practices
  * [PR #1234] minimal support for ros drones
  * [PR #1236] Added printable pieces for PiBot
  * [PR #1240] 3DVizWeb show objects
  * [PR #1241] Added new formula1 models
  * [PR #1250] Added Nurburgrin circuit
  * [PR #1251] solved several problems with pibot api
  * [PR #1256] velocity implemented with five speeds for pibot real driver.
  * [PR #1258] velocity implemented at "avanzar" function of piBot.py for real PiBot
  * [PR #1262] Added Cameras and no-laser models to the F1 models. Follow_Road world fixed
  * [PR #1271] Control in position for pibot driver
  * [PR #1277] drone model with mavros

-- Francisco Perez <f.perez475@gmail.com>, Tue, 20 Nov 2018 12:00:00 +0100

jderobot (5.6.4) unstable; urgency=low

  * release 5.6.4
  * Solved some bugs related with installation and other minor bugs in components.
  * Added some little improvements in some tools. 
  * [PR #1070] Now server value on yml files are more intuitive.
  * [PR #1080] Change in movement blocks and added math blocks
  * [PR #1082] Added support for kinect1's ROS driver
  * [PR #1091] HSV color disk added to colorTuner tool
  * [PR #1106] Added support for kinect2's ROS driver  
  * [PR #1111] Moved scratch2jderobot away from the oficial repo. Now this tool has its own one.
  * [PR #1119] Added turtlebot ROS launchers
  * [PR #1123] Added 3dViewer-web tool
  * [PR #1124] New tool cameraview-web
  * [PR #1125] Deleted 3dViewer tool.
  * [PR #1126] Cameraserver python full ROS based and full OpenCV
  * [PR #1127] Added new world for IIROS drone championship
  * [PR #1132] Better support for bumper sensor in comm
  * [PR #1133] Web tool kobukiviewer renamed and electron based  
  * [PR #1134] Web tool uavviewer renamed and electron based
  * [PR #1141] Added new ROS based formula1 model for simulation
  * [PR #1148] Renamed several components 
  * [PR #1150] New models for holo taxi and opel car ROS based.
  * [PR #1154] Deleted old webtools 

-- Francisco Perez <f.perez475@gmail.com>, Wed, 18 May 2018 14:20:00 +0100

jderobot (5.6.3) unstable; urgency=low

  * release 5.6.3
  * Solved some bugs related with installation and other minor bugs in components.
  * Added some little improvements in some tools. 
  * YoutubeServer is now working from debian package installation, it only works with python clients for the moment.
  * [PR #1017] Added support for synchronized RGBD
  * [PR #1024] Added new tests to jenkins dockers
  * [PR #1031] Added new component rgbdViewer_py
  * [PR #1032] Added support for MBOT in gazebo 
  * [PR #1040] Added a new widget with zoom to rgbdcalibrator
  * [PR #1049] New tests for packages installation.
  * [PR #1050] First approach. Deleted discontinuity in HSV filter of colorTuner
  * [PR #1053] Added comm to rgbdViewer 
  * [PR #1054] Ported MAVLINKServer to python 2.7 and integrated with comm 
  * [PR #1035] [PR #1063] Improved scratch2JdeRobot tool.

-- Francisco Perez <f.perez475@gmail.com>, Wed, 19 Feb 2018 17:15:00 +0100


jderobot (5.6.2) unstable; urgency=low

  * release 5.6.2
  * Solved some bugs and updated to even ROS update packages due to opencv package stopped working.
  * Added some little improvements in some tools. 
  * [PR #963] Scratch2JdeRobot: generation of output node in a single step
  * [PR #966] Added suppor for bumper sensor in comm.
  * [PR #969] Added new server for cameras in python using openCV instead our visionlib
  * [PR #973] Added examples for visualStates
  * [PR #974] Improvements in visualStates.
  * [PR #975] Added new interface for pause, stop, resume, etc.
  * [PR #979] Webtools are operative again.
  * [PR #982] Migrated uav_viewer to make it compatible with python 2.7
  * [PR #988] Deleted kobuki_driver and laser_server and added ros drivers for kobuki and lasers.
  * [PR #950] [PR #1001] Added new server: Youtube server. Now we can have another source of video.
  * [PR #1006] Final and fully functional version of visualStates.


scripts/install_deps.sh
#!/bin/bash

sudo sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
sudo apt-key adv --keyserver 'hkp://keyserver.ubuntu.com:80' --recv-key C1CF6E31E6BADE8868B172B4F42ED6FBAB17C654

sudo sh -c 'echo "deb http://packages.osrfoundation.org/gazebo/ubuntu-stable `lsb_release -cs` main" > /etc/apt/sources.list.d/gazebo-stable.list'
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-key 67170598AF249743

sudo apt-key adv --keyserver keyserver.ubuntu.com --recv B6391CB2CFBA643D
sudo apt-add-repository "deb http://zeroc.com/download/Ice/3.7/ubuntu18.04 stable main"

sudo sh -c 'cat<<EOF>/etc/apt/sources.list.d/jderobot.list
# for ubuntu 16.04 LTS (64 bit)

deb [arch=amd64] http://jderobot.org/apt bionic main
EOF'
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv 24E521A4

sudo apt update

sudo apt install wget

sudo apt install -y ros-melodic-roscpp ros-melodic-std-msgs ros-melodic-cv-bridge ros-melodic-image-transport ros-melodic-roscpp-core ros-melodic-rospy ros-melodic-nav-msgs ros-melodic-geometry-msgs ros-melodic-kobuki-msgs
sudo apt-get install -y libpcl-dev
sudo apt-get install libopencv-dev
sudo apt install -y build-essential libtool cmake g++ gcc git make

sudo apt install -y freeglut3 freeglut3-dev libgl1-mesa-dev libglu1-mesa

sudo apt-get install -y libgtk2.0-0 libgtk2.0-bin libgtk2.0-cil libgtk2.0-common libgtk2.0-dev libgtkgl2.0-1
sudo apt-get install -y libgtkgl2.0-dev libgtkglext1 libgtkglext1-dev libglademm-2.4-dev libgtkmm-2.4-dev 
sudo apt-get install -y libgnomecanvas2-0 libgnomecanvas2-dev  libgtkglext1-doc libgnomecanvasmm-2.6-dev
sudo apt-get install -y libgnomecanvasmm-2.6-1v5 libgtkglextmm-x11-1.2-0v5 libgtkglextmm-x11-1.2-dev

sudo apt-get install -y libgoocanvasmm-2.0-6 libgoocanvasmm-2.0-dev

sudo apt-get install -y libgsl23 gsl-bin libgsl-dev

sudo apt-get install -y libxml++2.6-2v5 libxml++2.6-dev libtinyxml-dev

sudo apt-get install -y libeigen3-dev

sudo apt-get install -y libdc1394-22 libdc1394-22-dev

sudo apt-get install -y libusb-1.0-0 libusb-1.0-0-dev

sudo apt-get install -y python-matplotlib python-pyqt5 python-pip python-numpy python-pyqt5.qtsvg python-pyqt5.qsci

sudo apt-get install -y qfi

sudo apt-get install -y qtbase5-dev libqt5script5 libqt5svg5-dev

sudo apt-get install -y libboost-system-dev libboost-filesystem-dev

sudo apt-get install -y libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev

sudo apt install -y libzeroc-ice3.7 libzeroc-icestorm3.7 zeroc-ice-slice libzeroc-ice-dev

sudo apt-get install -y libssl-dev libbz2-dev
sudo pip2 install --upgrade pip
sudo pip2 install zeroc-ice==3.6.4

sudo apt-get install -y libopenni2-dev libopenni-dev

sudo apt-get install -y gazebo9 libgazebo9-dev

sudo apt-get install -y nodejs

sudo apt-get install -y ardronelib libgoogle-glog-dev

sudo apt install libyaml-cpp-dev








